### PR TITLE
[POC] Support of modern controllers on module tabs

### PR DESCRIPTION
**Pull request by mickaelandrieu**
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | Allow modern controllers to be registered using Module::$tabs property.
| Type?         | improvement
| Category?     | BO
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? | dunno
| How to test?  | it's a POC, I won't have time to finish it (and unregistration may be updated accordingly?)

ping @Quetzacoalt91 for review ;)

### Use

```php
   $this->tabs = array(
        array(
            'name' => 'Yolo',
            'class_name' => 'adminmessageoftheday', // legacy
            'visible' => true,
            'parent_class_name' => 'AdminTraining',
        ),
        array(
            'name' => 'Yolo2',
            'class_name' => ExempleController::class,
            'legacy_class_name' => 'AdminTrainingIndexClass',
            'visible' => true,
            'parent_class_name' => 'AdminTraining',
        ));
```

```yml
admin_ps_training_index:
  path: /training/
  methods: [GET]
  defaults:
    _controller: 'PrestaShop\Training\Controller\Admin\ExempleController::indexAction'
    _legacy_controller: 'AdminTrainingIndexClass'
    _legacy_link: 'AdminTrainingIndexClass'
```

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/13889)

### CODE REVIEW

File: src/Adapter/Module/Tab/ModuleTabRegister.php [hunk 1]
```diff
@@ -275,15 +282,18 @@ protected function registerTab(Module $module, ParameterBag $tabDetails)
     {
         $this->checkIsValid($module->get('name'), $tabDetails);
 
-        /**
-         * Legacy Tab, to be replaced with Doctrine entity when right management
-         * won't be directly linked to the tab creation
-         *
-         * @ToDo
-         */
+        // Legacy Tab, to be replaced with Doctrine entity when right management
```
Comment (mickaelandrieu) [inline src/Adapter/Module/Tab/ModuleTabRegister.php:285]: maybe this comment/todo can be dropped now?

File: src/Adapter/Module/Tab/ModuleTabRegister.php [hunk 2]
```diff
@@ -275,15 +282,18 @@ protected function registerTab(Module $module, ParameterBag $tabDetails)
...
+        // won't be directly linked to the tab creation
+        // @ToDo
+
+        $className = $tabDetails->get('class_name');
+
+        if (is_subclass_of($className, FrameworkBundleAdminController::class)) {
```
Comment (Quetzacoalt91) [inline src/Adapter/Module/Tab/ModuleTabRegister.php:291]: What's the difference between this and

```
         if ($className instanceof FrameworkBundleAdminController) { 
```

File: src/Adapter/Module/Tab/ModuleTabRegister.php [hunk 3]
```diff
@@ -275,15 +282,18 @@ protected function registerTab(Module $module, ParameterBag $tabDetails)
...
+            $className = $tabDetails->get('legacy_class_name');
```
Comment (mickaelandrieu) [inline src/Adapter/Module/Tab/ModuleTabRegister.php:292]: thinking about it, it's a very naive implementation: I should check for the existence of this parameter and throw an exception with a meaningful message pointing to the docs: wdyt?

File: src/Adapter/Module/Tab/ModuleTabRegister.php (hunk 2 above)
Comment (PierreRambaud) [reply src/Adapter/Module/Tab/ModuleTabRegister.php:291]: An example si better than a long story: https://www.php.net/manual/en/internals2.opcodes.instanceof.php#109108
But don't know why you use `is_subclass_of` here :thinking:

Comment (Quetzacoalt91) [reply src/Adapter/Module/Tab/ModuleTabRegister.php:291]: Yeah I know the difference, my question was actually the same as yours. Is there a specific need here?

Comment (mickaelandrieu) [reply src/Adapter/Module/Tab/ModuleTabRegister.php:291]: if I remember well, `instanceof` only accepts instances and not class names: if I'm wrong, I'm also in favor of using `instanceof` instead.

> [instanceof](https://www.php.net/manual/fr/language.operators.type.php) est utilisé pour déterminer si une variable PHP est un objet instancié d'une certaine classe

**Comment by mickaelandrieu** [comment]
Closed in favor of https://github.com/PrestaShop/PrestaShop/issues/12640

**Comment by mickaelandrieu** [comment]
~~ping @jolelievre : as this issue is still not fixed, would you consider this contribution ?~~

Just noticed you have worked on https://github.com/PrestaShop/PrestaShop/pull/14679 👍

**Comment by jolelievre** [comment]
Yes, I was about to tell you but you found it :)

**Pull request closed without merging**